File: p040.java
/* 
 * Solution to Project Euler problem 40
 * By Nayuki Minase
 * 
 * http://nayuki.eigenstate.org/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */


public final class p040 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p040().run());
	}
	
	
	public String run() {
		StringBuilder sb = new StringBuilder();
		for (int i = 1; i < 1000000; i++)
			sb.append(i);
		
		int prod = 1;
		for (int i = 0; i <= 6; i++)
			prod *= sb.charAt(Library.pow(10, i) - 1) - '0';
		return Integer.toString(prod);
	}
	
}

File: p040.hs
{- 
 - Solution to Project Euler problem 40
 - By Nayuki Minase
 - 
 - http://nayuki.eigenstate.org/page/project-euler-solutions
 - https://github.com/nayuki/Project-Euler-solutions
 -}


digits 0 = [0]
digits n = reverse (digits' n)
digits' 0 = []
digits' n = (mod n 10) : (digits' (div n 10))

decimal = foldr (\n rest -> digits n ++ rest) [] [1..]

ans = product [decimal !! (10^i - 1) | i <- [0..6]]

main = putStrLn (show ans)
